### PR TITLE
feat(plugins): add gateway-backed subagent runtime dispatch

### PR DESCRIPTION
**Pull request by livingghost**
Supersedes #56527.

## Summary

- Problem: Gateway-backed plugins need a stable `runtime.subagent` binding, but the current runtime path does not consistently inherit request-scoped or fallback runtime options, and can tag the active registry with the wrong runtime mode.
- Why it matters: Without a dedicated gateway-backed runtime binding, plugins can miss subagent capabilities, cache the wrong runtime shape, or dispatch with the wrong synthetic scopes.
- What changed: Added a shared gateway-backed subagent runtime, taught plugin loading to inherit shared runtime options only when no explicit runtime options are supplied, kept the loader/runtime cache keys sensitive to those capabilities, derived the active runtime mode from the effective runtime options, and rolled back the installed gateway runtime if plugin loading fails.
- What did NOT change (scope boundary): This PR does not add structured-output or tool-control forwarding, and it does not expose `subagent.enqueue()` or `subagent.abort()` because those gateway handlers do not exist on main.

## Change Type (select all)

- [ ] Bug fix
- [x] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [x] API / contracts
- [x] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #
- Related #
- [ ] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: N/A
- Missing detection / guardrail: N/A
- Prior context (`git blame`, prior PR, issue, or refactor if known): Split from `fork/feat/subagent-enhancements`
- Why this regressed now: N/A
- If unknown, what was ruled out: N/A

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [ ] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `src/gateway/server-plugins.test.ts`, `src/plugins/runtime/index.test.ts`, `src/agents/runtime-plugins.test.ts`, `src/plugins/tools.optional.test.ts`, `src/plugins/loader.test.ts`
- Scenario the test should lock in: gateway-backed runtime binding installs the right synthetic scopes, loader/runtime cache does not erase subagent runtime capabilities, failed plugin loads do not leak the temporary gateway runtime, explicit runtime-option call sites do not pretend to inherit shared runtime options, inherited gateway binding tags the active registry as `gateway-bindable`, and explicit `allowGatewaySubagentBinding: false` remains a real opt-out.
- Why this is the smallest reliable guardrail: the branch changes the gateway/runtime/loader seam rather than a single isolated helper.
- Existing test that already covers this (if any): the named suites already existed and were extended or aligned here.
- If no new test is added, why not: N/A

## User-visible / Behavior Changes

- `runtime.subagent` on this branch exposes `run`, `waitForRun`, and session accessors through the gateway-backed runtime.
- Fallback session reads use a synthetic read scope; runs and waits use synthetic write scope.
- Plugin loader cache keys now reflect inherited runtime capabilities, so a runtime with subagent support is not confused with one without it.
- Failed gateway plugin loads restore the previous shared runtime options and gateway subagent runtime before rethrowing.

## FAQ / Known review points

- Q: Why does this PR not expose `subagent.enqueue()` or `subagent.abort()`?
  - A: Those gateway handlers do not exist on main, so this split keeps the runtime surface to the methods that are actually backed by gateway behavior: `run`, `waitForRun`, and session accessors.
- Q: Why are read and write synthetic scopes split?
  - A: Session reads use synthetic read scope, while run and wait operations use synthetic write scope, matching the intended gateway authorization boundary.
- Q: Why is `inheritSharedRuntimeOptions` not passed when explicit gateway binding is enabled?
  - A: Once a caller passes explicit `runtimeOptions`, the loader uses that object directly. The branch only opts into shared runtime inheritance on call sites that do not provide explicit runtime options.
- Q: Why does shared runtime state store `allowGatewaySubagentBinding: true` instead of a concrete subagent object?
  - A: Shared state needs to preserve late binding to the process-global gateway runtime. Storing the flag keeps `clearGatewaySubagentRuntime()` authoritative and avoids leaking a stale concrete subagent across reloads.
- Q: What happens if plugin loading fails after installing the gateway-backed runtime?
  - A: The previous shared runtime options and gateway subagent runtime are restored before rethrowing, so a failed load does not leak partial runtime state.
- Q: Why do loader/runtime cache keys change in this PR?
  - A: The cache key now captures inherited runtime capability metadata so plugins loaded with a gateway-backed subagent runtime are not reused in contexts that lack that capability.
- Q: Why does the active runtime mode derive from effective runtime options instead of only explicit `runtimeOptions`?
  - A: Inherited shared runtime options can also enable gateway subagent binding. The active registry tag has to match the effective runtime shape, or gateway-bindable registries get mislabeled as `default` and miss the reuse fast path.
- Q: Why does `runtime.subagent.run()` synthesize an idempotency key when the caller omits one?
  - A: The gateway `agent` schema requires a non-empty `idempotencyKey`. Omitting it now mints a fresh per-call key so the dispatch still produces a distinct run rather than failing validation.
- Q: What does `allowGatewaySubagentBinding: false` do on this branch?
  - A: Explicit `false` stays a real opt-out. Only `undefined` inherits shared runtime options; `false` passes neither explicit runtime options nor shared runtime inheritance.

## Diagram (if applicable)

```text
Before:
[gateway plugin load]
  -> [runtime created without shared subagent binding]
  -> [subagent methods missing or cached incorrectly]

After:
[gateway plugin load]
  -> [shared gateway subagent runtime installed]
  -> [loader/runtime inherit capability key and mode]
  -> [run/wait/session access dispatch with correct synthetic scopes]
```

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) Yes
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) Yes
- Data access scope changed? (`Yes/No`) Yes
- If any `Yes`, explain risk + mitigation:
  - The runtime now exposes gateway-backed subagent operations through a late-bound runtime. Mitigation is explicit scope separation in `server-plugins.ts` (`read` for session reads, `write` for run and wait) plus existing gateway authorization.

### Environment

- OS: Windows 11
- Runtime/container: local `pnpm` workspace
- Model/provider: N/A
- Integration/channel (if any): gateway/plugin runtime only
- Relevant config (redacted): default repo config

### Steps

1. Run `pnpm check`.
2. Run `pnpm build:plugin-sdk:dts`.
3. Run `pnpm plugin-sdk:api:check`.
4. Run `pnpm exec oxfmt --check src/agents/runtime-plugins.ts src/agents/runtime-plugins.test.ts src/auto-reply/templating.ts src/gateway/server-plugins.ts src/gateway/server-plugins.test.ts src/plugins/loader.ts src/plugins/loader.test.ts src/plugins/runtime/index.ts src/plugins/runtime/shared-runtime-options.ts src/plugins/runtime/types.ts src/plugins/status.ts src/plugins/tools.ts src/plugins/tools.optional.test.ts src/plugins/types.ts`.
5. Run `pnpm exec vitest run --configLoader native --pool=threads src/gateway/server-plugins.test.ts src/plugins/runtime/index.test.ts src/agents/runtime-plugins.test.ts src/plugins/tools.optional.test.ts src/plugins/loader.test.ts`.

### Expected

- Type, lint, and format checks pass.
- Plugin-sdk build and baseline checks pass.
- Gateway runtime binding, loader cache-key, runtime mode, and runtime proxy tests pass.

### Actual

- All of the above passed locally on this rebuild branch based on latest `upstream/main`.

## Evidence

- [x] Failing test/log before + passing after
- [ ] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

## Human Verification (required)

- Verified scenarios: `pnpm check`, plugin-sdk type and baseline checks, format checks, and targeted gateway/runtime tests.
- Edge cases checked: fallback session reads use read scope, failed gateway plugin loads restore the previous runtime binding, explicit runtime-option call sites do not carry dead shared-runtime inheritance, inherited gateway binding marks active registries as `gateway-bindable`, explicit `allowGatewaySubagentBinding: false` does not fall through to shared runtime inheritance, and empty registry defaults stay intact.
- What you did **not** verify: full end-to-end plugin startup beyond the changed gateway/runtime seam.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps:

## Risks and Mitigations

- Risk: shared runtime options could leak the wrong runtime shape into the loader cache.
  - Mitigation: cache keys now include runtime capability metadata, the active runtime mode now derives from the effective runtime options, and the targeted loader/runtime tests passed.
- Risk: gateway runtime state could leak across failed startup attempts.
  - Mitigation: failed plugin loads now restore the previous shared runtime options and gateway subagent runtime before rethrowing.

### CODE REVIEW

File: src/gateway/server-plugins.ts [hunk 1]
```diff
@@ -361,13 +416,20 @@ export function createGatewaySubagentRuntime(): PluginRuntime["subagent"] {
       return {
         status,
         ...(typeof payload?.error === "string" && payload.error && { error: payload.error }),
+        ...(typeof payload?.stopReason === "string" &&
+          payload.stopReason && { stopReason: payload.stopReason }),
+        ...(Array.isArray(payload?.pendingToolCalls) && payload.pendingToolCalls.length > 0
+          ? { pendingToolCalls: payload.pendingToolCalls }
+          : {}),
       };
```
[the remark this answers is not in the document]
Comment (livingghost) [reply src/gateway/server-plugins.ts:448]: Addressed in cfca5e9f06. waitForRun now validates pendingToolCalls item shape at runtime and drops malformed arrays instead of passing them through.

File: src/plugins/loader.ts [hunk 2]
```diff
@@ -1,1714 +1 @@
...
...nue;     }      const api = buildPluginApi({       id: record.id,       name: record.name,       version: record.version,       description: record.description,       source: record.source,       rootDir: record.rootDir,       registrationMode: "cli-metadata",       config: cfg,       pluginConfig: validatedConfig.value,       runtime: {} as PluginRuntime,       logger,       resolvePath: (input) => resolveUserPath(input),       handlers: {         registerCli: (registrar, opts) => registerCli(record, registrar, opts),       },     });      try {       await register(api);       registry.plugins.push(record);       seenIds.set(pluginId, candidate.origin);     } catch (err) {       recordPluginError({         logger,         registry,         record,         seenIds,         pluginId,         origin: candidate.origin,         error: err,         logPrefix: `[plugins] ${record.id} failed during register from ${record.source}: `,         diagnosticMessagePrefix: "plugin failed during register: ",       });     }   }    return registry; }  function safeRealpathOrResolve(value: string): string {   try {     return fs.realpathSync(value);   } catch {     return path.resolve(value);   } }
```
[the remark this answers is not in the document]
Comment (livingghost) [reply src/plugins/loader.ts]: You were right - this was a bad branch republish on my side. The remote branch briefly carried a corrupted one-line version of `src/plugins/loader.ts`. I have force-published the clean local head (`2c0d83108c`) and verified that the file on GitHub is back to its normal multiline form. Sorry about the noise here, and thanks for flagging it.

File: src/gateway/server-plugins.ts [hunk 3]
```diff
@@ -434,24 +518,46 @@ export function loadGatewayPlugins(params: {
...
-      allowGatewaySubagentBinding: true,
-    },
-    preferSetupRuntimeForChannelPlugins: params.preferSetupRuntimeForChannelPlugins,
+  const previousGatewaySubagentRuntime = getGatewaySubagentRuntime();
+  const gatewaySubagentRuntime = createGatewaySubagentRuntime();
+  setGatewaySubagentRuntime(gatewaySubagentRuntime);
+  const previousSharedRuntimeOptions = getSharedPluginRuntimeOptions();
+  setSharedPluginRuntimeOptions({
+    allowGatewaySubagentBinding: true,
   });
```
[the remark this answers is not in the document]
Comment (livingghost) [reply src/gateway/server-plugins.ts]: Initialized the shared runtime options before the empty-plugin early return so gateway startup still publishes the binding contract when `pluginIds` is empty. Updated the existing test to exercise the zero-plugin path.
